### PR TITLE
Allow manually creating custom aliases

### PR DESCRIPTION
**Pull request by Vinnl**
# New feature description

This allows users to create custom aliases themselves, without having to send an email to it.

# Screenshot (if applicable)

How about a video?

https://user-images.githubusercontent.com/4251/150332894-7319960d-621b-412e-88b3-4b6645aa3e49.mp4

# How to test

1. Log in using an account that has Premium _and_ has a custom subdomain.
2. Click "Generate new alias".
3. Verify that a menu pops up, through which you can create your own custom alias.
4. Also make sure that a non-Premium user, or a Premium user without a custom subdomain, just has a regular button that directly creates a random alias.

# Checklist

- [x] l10n dependencies have been merged, if any. - Most of them, but I forgot two unfortunately: https://github.com/mozilla-l10n/fx-private-relay-l10n/pull/55
- [ ] All acceptance criteria are met. - It also says it should shop up on top, but alias sorting I think is a separate PR that might conflict with https://github.com/mozilla/fx-private-relay-add-on/pull/242, so I left that for a potential follow-up PR.
- [x] All UI revisions follow the [coding standards](https://github.com/mozilla/fx-private-relay/blob/main/docs/coding-standards.md), and use Protocol tokens where applicable (see `/static/scss/libs/protocol/css/includes/tokens/dist/index.scss`).
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -2,4 +2,5 @@ node_modules
 static/js/clipboard.min.js
 venv
 env
-static/scss/libs
\ No newline at end of file
+static/scss/libs
+frontend/out
```
Comment (Vinnl) [inline .eslintignore:6]: I'm sorry, I snuck in an unrelated change, but this directly might be present if you've checked out the `react-experiment` branch in the past, and it's annoying if ESLint tries to lint it.

**Review by maxxcrawford — CHANGES_REQUESTED** [review]
Nice feature build-out! I'm excited to see this get built out. 

- Keyboard accessibility notes: 
  - Add :focus/hover state for sub-items when the alias create panel is open (from keyboard only)
  - When the modal opens, you may focus on that input

Blocking concern: When I tried to create the same alias twice, I did not get any error state/messaging — it just didn't work/submit. Can you add sort of error state?

File: privaterelay/templates/includes/dashboard-filter.html [hunk 2]
```diff
@@ -3,6 +3,8 @@
 {% load ftl %}
 {% ftlconf bundle='privaterelay.ftl_bundles.main' %}
 
+<script src="{% static 'js/custom-alias-picker.js' %}"></script>
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply privaterelay/templates/includes/dashboard-filter.html]: Excellent idea: ba6df120

File: privaterelay/templates/includes/dashboard-filter.html [hunk 3]
```diff
@@ -126,27 +128,60 @@
...
                     >
                         <span class="generate-new-relay-icon"></span>
                         <span class="generate-new-alias-text">
                             {% ftlmsg 'profile-label-generate-new-alias' %}
                         </span>
                     </button>
                 {% endif %}
+            {% elif not user_profile.subdomain %}
+                <button
+                    class="blue-btn-states dash-create-new-relay flx al-cntr jst-cntr"
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply privaterelay/templates/includes/dashboard-filter.html]: 075975fa

File: privaterelay/templates/includes/dashboard-filter.html [hunk 4]
```diff
@@ -126,27 +128,60 @@
...
+                    <button
+                        class="js-dash-create-new-alias-menu-trigger blue-btn-states dash-create-new-alias-menu-trigger flx al-cntr jst-cntr"
+                        title="{% ftlmsg 'profile-label-generate-new-alias' %}"
+                    >
+                        <span class="generate-new-alias-text">
+                            {% ftlmsg 'profile-label-generate-new-alias' %}
+                        </span>
+                        <span class="generate-new-alias-menu-trigger-icon"></span>
+                    </button>
+                    <div class="js-dash-create-new-alias-menu-popup generate-new-alias-menu-popup hidden">
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply privaterelay/templates/includes/dashboard-filter.html]: 4f2542fb

File: static/js/app.js [hunk 5]
```diff
@@ -236,22 +236,30 @@ function addEventListeners() {
...
-  const moveDownBy = parseInt(setMenuLinksHeightVal, 10);
+  const generateAliasMenuTrigger = document.querySelector(".js-dash-create-new-alias-menu-trigger ");
+  const generateAliasMenuPopup = document.querySelector(".js-dash-create-new-alias-menu-popup ");
 
-  if (!mainWrapper.style.marginTop || mainWrapper.style.marginTop === "0px") {
-    mainWrapper.style.marginTop = moveDownBy + "px";
-  }
+  generateAliasMenuTrigger.addEventListener("click", (e) => {
+    e.preventDefault();
+    generateAliasMenuPopup.classList.toggle("hidden");
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply static/js/app.js]: 4f2542fb

File: static/js/app.js [hunk 6]
```diff
@@ -236,22 +236,30 @@ function addEventListeners() {
...
+    const moveDownBy = parseInt(setMenuLinksHeightVal, 10);
+
+    if (!mainWrapper.style.marginTop || mainWrapper.style.marginTop === "0px") {
+      mainWrapper.style.marginTop = moveDownBy + "px";
+    }
 
-  else {
-    mainWrapper.style.marginTop = 0;
+    else {
+      mainWrapper.style.marginTop = 0;
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply static/js/app.js:256]: I didn't change this code; merely fixed the indentation. It was introduced in #1326. That said, I assume it's the same as the difference between

```css
margin-top: 0;
```

and

```css
margin-top: 0px
```

which are equivalent, and for which we don't have a rule in stylelint mandating one or the other. So for now, I think the answer would be: not necessarily :)

File: static/js/custom-alias-picker.js [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+                    return;
+                }   
+                
+                const modal = document.querySelector(".js-modal-custom-alias-picker");
+                modal.classList.remove("is-visible");
+
+                document.removeEventListener("keydown", customAliasPicker.modal.close, false);
+
+            },
+            formSubmitRequest: async (_e) => {
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply static/js/custom-alias-picker.js]: Nope (hence the `_` prefix), I left it in as a nod that it's a regular submit event listener, but might as well remove it: f9185f93

File: static/scss/partials/main.scss [hunk 8]
```diff
@@ -568,14 +568,76 @@ main, table {
...
 .dash-create-new-relay.btn-disabled {
     background-color: $color-grey-30
 }
 
+.dash-create-new-alias-menu-wrapper {
+    position: relative;
+}
+
+.dash-create-new-alias-menu-trigger {
+    border-radius: 4px;
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply static/scss/partials/main.scss]: 7e91b12a

File: static/scss/partials/main.scss [hunk 9]
```diff
@@ -568,14 +568,76 @@ main, table {
...
+    padding: 0 .25em 0 .75em;
+    margin-left: 1em;
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply static/scss/partials/main.scss]: fa41c08d

**Comment by Vinnl** [comment]
@maxxcrawford Whoops, I did actually add an error state, but called a non-existent method (`customAliasPicker.modal.showError` rather than `customAliasPicker.showError`). If only there was a tool that would warn me of such things! :P 75bfc4cc

I'm not sure what you mean about adding focus states though - I think they all already had one? See screencast below.

https://user-images.githubusercontent.com/4251/150819853-0e4c69c6-3ed2-4dcf-a408-a94626c5b5a4.mp4

File: static/js/custom-alias-picker.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply static/js/custom-alias-picker.js]: Got'cha! That nod makes sense.

**Review by maxxcrawford — APPROVED** [review]
🏁 Off to the races!

**Pull request merged**